src/python/arcor2_arserver/rpc/project.py

```diff
@@ -605,26 +660,30 @@ async def update_action_point_orientation_using_robot_cb(
     :return:
     """
 
-    ensure_scene_started()
+    async with ctx_read_lock(req.args.robot.robot_id, "ui.owner"):
+        ensure_scene_started()
 
-    assert glob.SCENE
-    assert glob.PROJECT
+        assert glob.SCENE
+        assert glob.PROJECT
 
-    ap, ori = glob.PROJECT.bare_ap_and_orientation(req.args.orientation_id)
+        ap, ori = glob.PROJECT.bare_ap_and_orientation(req.args.orientation_id)
 
-    new_pose = await get_end_effector_pose(req.args.robot.robot_id, req.args.robot.end_effector)
+        if not await glob.LOCK.is_write_locked(ap.id, "ui.owner"):
+            raise LockingException("Action point needs to be locked before manipulating")
 
-    if ap.parent:
-        new_pose = tr.make_pose_rel_to_parent(glob.SCENE, glob.PROJECT, new_pose, ap.parent)
+        new_pose = await get_end_effector_pose(req.args.robot.robot_id, req.args.robot.end_effector)
 
-    ori.orientation = new_pose.orientation
+        if ap.parent:
+            new_pose = tr.make_pose_rel_to_parent(glob.SCENE, glob.PROJECT, new_pose, ap.parent)
 
-    glob.PROJECT.update_modified()
+        ori.orientation = new_pose.orientation
 
-    evt = sevts.p.OrientationChanged(ori)
-    evt.change_type = Event.Type.UPDATE
-    asyncio.ensure_future(notif.broadcast_event(evt))
-    return None
+        glob.PROJECT.update_modified()
+
+        evt = sevts.p.OrientationChanged(ori)
```

Comment:
Přemýšlím jestli by nebylo lepší tyhle ty části kódu kde se posílají události už mimo zámek. Co myslíte? Kdyby to mělo být 100% neprůstřelné, asi by se těsně před opuštěním context manageru musela udělat kopie toho `ori` (v tomto případě), aby před odesláním události nemohlo dojít v jiném callbacku ke změně.

Reply:
No, ono se to vlastně reálně posílá už teď mimo zámek, protože odeslání je "odložené" přes `ensure_future`. Čili přesunutím té části kódu až za context manager by se toho vlastně moc nezměnilo (z pohledu rizika nežádoucích modifikací). Tak možná to přesunout? A zvážit vytvoření té kopie dat co se posílají v rámci události?

Reply:
Já v tom nevidím úplně rozdíl. Zkoušel jsem navodit stav, kdy by se "vetřel" task mezi ukončení kontext manažera a pokračování zbytku funkce, ale pokud tam není žádné asynchronní volání, tak ta funkce vždy pokračovala bez přerušení... V tomhle případě si dokonce myslím, že naplánovaný task bude označen jako splnitelný a předběhne nový, který může uživatel vytvořit mezi ukončením zámku a notifikací.

Reply:
Reálně to asi bude tak jak říkáte, že nejdřív proběhne ten task s odesláním notifikace. Ale garantované to nijak není.